File: apps/backend/app/models/task.py
import uuid
import enum
from sqlalchemy import Column, String, DateTime, ForeignKey, Text, Enum, Integer
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.db.session import Base


class TaskStatus(str, enum.Enum):
    todo = "todo"
    doing = "doing"
    done = "done"


class TaskPriority(str, enum.Enum):
    low = "low"
    medium = "medium"
    high = "high"
    urgent = "urgent"


class Task(Base):
    __tablename__ = "tasks"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id = Column(String, ForeignKey("projects.id"), nullable=False)
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    due_date = Column(DateTime(timezone=True), nullable=True)
    status = Column(Enum(TaskStatus), default=TaskStatus.todo)
    priority = Column(Enum(TaskPriority), default=TaskPriority.medium)
    assignee_id = Column(String, ForeignKey("users.id"), nullable=True)
    position = Column(Integer, default=0)
    created_by = Column(String, ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    project = relationship("Project", back_populates="tasks")
    assignee = relationship("User", foreign_keys=[assignee_id])
    creator = relationship("User", foreign_keys=[created_by])
    comments = relationship("Comment", back_populates="task", cascade="all, delete-orphan")
    attachments = relationship("Attachment", back_populates="task", cascade="all, delete-orphan")
    time_entries = relationship("TimeEntry", back_populates="task", cascade="all, delete-orphan")

File: apps/frontend/src/types/task.ts
// src/types/task.ts
export type TaskStatus = 'todo' | 'in-progress' | 'review' | 'done';
export type TaskPriority = 'low' | 'medium' | 'high' | 'urgent';

export interface Task {
  id: string;
  title: string;
  description: string;
  status: TaskStatus;
  priority: TaskPriority;
  dueDate?: string;
  projectId: string;
  assigneeId?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateTaskData {
  title: string;
  description: string;
  status: TaskStatus;
  priority: TaskPriority;
  dueDate?: string;
  projectId: string;
}

export interface UpdateTaskData {
  title?: string;
  description?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  dueDate?: string;
  projectId?: string;
}

export interface Project {
  id: string;
  name: string;
  color: string;
  status: 'active' | 'archived';
}

export interface User {
  id: string;
  name: string;
  email: string;
}




// // types/task.ts
// export type TaskStatus = 'todo' | 'in-progress' | 'review' | 'done';
// export type TaskPriority = 'low' | 'medium' | 'high' | 'urgent';

// export interface Task {
//   id: string;
//   title: string;
//   description: string;
//   status: TaskStatus;
//   priority: TaskPriority;
//   dueDate?: string;
//   projectId: string;
//   assigneeId?: string;
//   createdAt: string;
//   updatedAt: string;
// }

// export interface CreateTaskData {
//   title: string;
//   description: string;
//   status: TaskStatus;
//   priority: TaskPriority;
//   dueDate?: string;
//   projectId: string;
// }

// export interface UpdateTaskData {
//   title?: string;
//   description?: string;
//   status?: TaskStatus;
//   priority?: TaskPriority;
//   dueDate?: string;
//   projectId?: string;
// }




// export interface Task {
//   id: string;
//   title: string;
//   description?: string;
//   projectId: string;
//   assignedTo?: string;
//   dueDate?: string;
//   priority: "low" | "medium" | "high";
//   status: "todo" | "in-progress" | "review" | "done";
//   position: number;
//   createdAt: string;
//   updatedAt: string;
//   labels?: string[];
// }

// export interface CreateTaskData {
//   title: string;
//   description?: string;
//   projectId: string;
//   assignedTo?: string;
//   dueDate?: string;
//   priority: "low" | "medium" | "high";
//   status: "todo" | "in-progress" | "review" | "done";
//   labels?: string[];
// }

// export interface UpdateTaskData {
//   title?: string;
//   description?: string;
//   assignedTo?: string;
//   dueDate?: string;
//   priority?: "low" | "medium" | "high";
//   status?: "todo" | "in-progress" | "review" | "done";
//   position?: number;
//   labels?: string[];
// }